### PR TITLE
Ensure <Nullable>enable<Nullable> on System.IO.Packaging and S.Resources.Extensions

### LINKED ISSUE #41696: Some projects aren't enabled for nullable
I did a pass to compare syntax of APIs as they are in reference assemblies and how they are in implementation assemblies. The full spreadsheet can be [downloaded from here](https://github.com/dotnet/runtime/files/5158462/nullable_differences_assembly_mismatch_only.zip). 

The spreadsheet only shows API differences if the reference assembly and the implementation assembly differ with their `<Nullable>enable</Nullable>` setting (i.e one defines `NullableAttribute` and one does not). This affects these assemblies:

* Microsoft.VisualBasic.Core
* System.CodeDom
* System.IO.Ports
* System.Runtime.CompilerServices.Unsafe
* System.Security.Permissions

The first one is expected because we decided to only annotate the reference assembly; the others aren't.

/cc @dotnet/nullablefc

### PR DESCRIPTION
**Pull request by krwq**
Fixes: https://github.com/dotnet/runtime/issues/41696

Looking at the XLS and conversation seems like this is all that's left to do there

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

**Comment by safern** [comment]
@krwq these are the only projects we we're missing to enable nullable that we're annotated. However the spreadsheet has some other differences in between ref and src. Are you going to tackle those as well?

**Review by safern — APPROVED** [review]

**Comment by krwq** [comment]
@safern I couldn't see any specific issues on the spreadsheet, the projects there didn't seem to have src annotated (except for perhaps few files from shared directory)

**Comment by buyaa-n** [comment]
Did not updated diff in namespaces
1. System.ComponentModel.Annotations by assuming we didn't annotated the src completely for a reason (not sure who annotated that, need to clarify)
2. System.Data.Common same reason, not sure if we want to make the src and ref same
3. System.ComponentModel.EventBasedAsync cannot even find this ref and src, which project are they?
4. Also src missing some annotation, might be on purpose
5. System.IO.Packaging @safern  said it annotated correctly, we missed to add enable on the ref
6. System.Resources.Extensions @safern  said it annotated correctly, we missed to add enable on the ref

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (buyaa-n) [inline src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:78]: This might need to be nullable, didn't understood why made non null, but if I change it to nullable the entire collection nullability need to be changed

File: src/libraries/System.Security.Cryptography.Pkcs/ref/System.Security.Cryptography.Pkcs.netcoreapp.cs
Comment (buyaa-n) [inline src/libraries/System.Security.Cryptography.Pkcs/ref/System.Security.Cryptography.Pkcs.netcoreapp.cs:158]: I wonder why the compiler didn't warn for `X509ExtensionCollection extensions = null` parameter

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (buyaa-n) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:146]: We had lots of conversation about Path nullability then ended up with making nullable https://github.com/dotnet/runtime/pull/528#discussion_r361762878

**Comment by safern** [comment]
> System.ComponentModel.Annotations by assuming we didn't annotated the src completely for a reason (not sure who annotated that, need to clarify)
> System.Data.Common same reason, not sure if we want to make the src and ref same

@roji do you recall why?

> System.ComponentModel.EventBasedAsync cannot even find this ref and src, which project are they?

https://github.com/dotnet/runtime/tree/master/src/libraries/System.ComponentModel.EventBasedAsync/src

Thanks, for working on this, @buyaa-n

**Comment by buyaa-n** [comment]
> > System.ComponentModel.EventBasedAsync cannot even find this ref and src, which project are they?
> 
> https://github.com/dotnet/runtime/tree/master/src/libraries/System.ComponentModel.EventBasedAsync/sr

I checked that, it doesn't have `BeginInvoke(object? sender,   ProgressChangedEventArgs! e, AsyncCallback callback, object object);` and `Invoke(object? sender,   ProgressChangedEventArgs! e);` methods, maybe that is generated for delegates, then i will just fix delegates signature

**Comment by safern** [comment]
Yeah the problem is: `public delegate void ProgressChangedEventHandler(object! sender, ProgressChangedEventArgs! e)`

**Comment by buyaa-n** [comment]
Fixed System.Data.Common diffs, for some of them source looks correct, but for some ref looks correct, so updated as whichever feels correct to me, please review. Now only 2 ns left, System.ComponentModel.Annotations and System.Data.OleDb, these are not within group 9 => by my understanding, we didn't intend to annotate their src completely so the diff might be OK CC @stephentoub,

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (stephentoub) [inline src/libraries/System.Data.Common/ref/System.Data.Common.cs:866]: This looks wrong:
https://github.com/dotnet/runtime/pull/41082
cc: @roji

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (stephentoub) [inline src/libraries/System.Data.Common/ref/System.Data.Common.cs:912]: Did @roji or someone else finish going through all of these APIs to annotate them appropriately?

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:78]: Just taking a quick look at the implementation, it looks like this should be `object?` with `DisallowNull`, and the implementation should be changed to remove the `!` from the `return index < 0 ? null! : _list[index].Value;`.  I'm not sure why it was suppressed like that.
cc: @tarekgh 
(and if it was done there, seems like the rest of the code should be quickly audited to make sure the same suppression wasn't used elsewhere in similar cases)

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (krwq) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs:912]: I don't believe so, I'll revert this change

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (krwq) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs:866]: I'll revert this

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (stephentoub) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs:912]: > I don't believe so

The next obvious question then is... can we?  Leaving some APIs as oblivious is unexpected.

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (krwq) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:78]: missed this feedback

Comment (krwq) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:78]: @stephentoub it seems this is implementation of `IDictionary<string, object>` if I changed the nullability of the base there will be conflict between DisallowNull and base. IMO we should change nullability to `object?` but to not annotate this as `DisallowNull` and just let it throw or add the attribute and suppress the warning... suggestions?

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (safern) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: I believe these depend on TypeConverters being annotated correctly. However that will be for 6.0.0, I will annotate these as part of that if we agree we shouldn't annotate them now.

Comment (krwq) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: @stephentoub theoretically we could but:
- @roji might have more context why this wasn't annotated in the first place
- I believe we're supposed to not annotate any more new code even if we left parts of the assembly not annotated and make sure we have time to flush any bugs related although this arguably could be treated as a bug

Comment (stephentoub) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: > I believe these depend on TypeConverters being annotated correctly.

Ok

> I believe we're supposed to not annotate any more new code even if we left parts of the assembly not annotated and make sure we have time to flush any bugs related although this arguably could be treated as a bug

Ok, but we should at least make sure this is tracked in some way. I didn't see any mention in Jeff's summary of work still to do that suggests we're tracking still unannotated portions of System.Data.

Comment (krwq) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: I've updated https://github.com/dotnet/runtime/issues/41720 to have this info

Comment (safern) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: > Ok, but we should at least make sure this is tracked in some way. I didn't see any mention in @jeffhandley's summary of work still to do that suggests we're tracking still unannotated portions of System.Data. (maybe I missed it?)

@roji and @jeffhandley let me know that was part of annotating TypeConverters, so I was considering that as part of my task.

Comment (buyaa-n) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: I thought these were disabled from XML dependency and by my understanding XML dependency for System.Data.Common is resolved, so its safe to remove it, thanks for reverting

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs]: > Just taking a quick look at the implementation, it looks like this should be object? with DisallowNull

The indexer allow nulls so the original fix should be ok. 
```C#
 public object this[string key] { get { throw null; } set { } }
```

basically the indexer allow null in the getter but never return null.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs]: looks I confused myself. I think @stephentoub was right. it should be `object?`. so you may ignore my previous comment.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs]: I am now wondering what `DisallowNull` will do? does it will prevent the setter from using null values? of so then this will be wrong.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs
Comment (tarekgh) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: >DisallowNull [](start = 9, length = 12)

I am now wondering what DisallowNull will do? does it will prevent the setter from using null values? of so then this will be wrong.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: > does it will prevent the setter from using null values

Yes, it'll warn if the value may be null. Nulls are allowed to be added as values? Then shouldn't Add take an `object?` rather than `object`? And implement the interface with a nullable TKey? I'm wondering why this setter is special

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (buyaa-n) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs]: Yes, it will warn if somebody set null value, from reading the API comments again, it should allow null, also it returns null when `index < 0`: `return index < 0 ? null! : _list[index].Value;`. So it need to be just nullable. That could affect the entire dictionary: to allow Add null and return nullable Values etc which seems not intended. I guess we just make it nullable and suppress the warning for the generic parent.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: the indexer can allow null values. null values for this collection means remove the entry matching the key. This is the behavior that OpenTelemetry suggested for such collection. only keys cannot be null. So we need to remove `DisallowNull` I guess.

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs]: for sure the indexer should allow null so it should be nullable. DisallowNull should be removed.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: @buyaa-n reminded me in other comment the indexer can return null too. anyway, I think it make sense having the indexer is nullable and not using DisallowNull.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: What's the behavior when you call `Add(key, null);`?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: It will add the (key, null) to the collection. this is different than the index behavior. Add always add and used only if anyone want to have null value.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: Then the interface implementation is annotated incorrectly. (And the difference between the indexer and Add behavior is very confusing.)

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: We discussed that in details in the design review and that what we concluded there. I agree we need to fix the interface annotation and I think that is what we are doing now.

Comment (buyaa-n) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: That is what i meant in my original comment https://github.com/dotnet/runtime/pull/41731#discussion_r482682472, making interface implementation allow nullable value need change for the entire dictionary. Which i was not sure if we want to do with this PR, if we want to udpate that with this PR i can do that

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: > I agree we need to fix the interface annotation and I think that is what we are doing now.

I mean the whole interface, not just the indexer. Anywhere TValue is `object` needs to be changed to `object?`, including in the definition.

> We discussed that in details in the design review and that what we concluded there

I think that's not a good outcome. The only difference between setting a value into a dictionary and adding that value into the dictionary should be overwrite behavior.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs:62]: > I think that's not a good outcome. The only difference between setting a value into a dictionary and adding that value into the dictionary should be overwrite behavior.

There is more details around that because of all interfaces we had in Activity objects which allow adding tags with null values.

> Which i was not sure if we want to do with this PR, if we want to udpate that with this PR i can do that

I think we can do it in this PR. now need a new PR for that. I am not expecting the changes will be big anyway.

**Comment by jeffhandley** [comment]
Once we have this finished up and merged into `master`, we'll submit it for Tactics Ask Mode approval for RC2:

- Use the port bot to port this to release/5.0-rc2 (with a comment of `/backport to release/5.0-rc2`)
- Fill out the PR template in the PR created by the bot
- Label it with `Servicing-consider`
- Then I'll send an email to Tactics to request approval
- I suspect we'll talk about it during the Tactics meeting on Tuesday

**Comment by jeffhandley** [comment]
Darn; false positive on that backport; I thought I could bury the syntax deep within a comment and be OK.  I'm going to close that port PR.

**Comment by safern** [comment]
@jeffhandley FYI: no need to close it, you can keep adding commits to this PR and once merged/ready, you run the command again and it will update the open PR for the backport and it will include the missing commits which is pretty handy 😄

**Comment by jeffhandley** [comment]
With that commit from @buyaa-n, we're now ready to start the backport.  Let's see if the bot knows to reopen the closed PR. 😜 

/backport to release/5.0-rc2

**Comment by tarekgh** [comment]
Thanks @buyaa-n. the changes in System.Diagnostic.DiagnosticsSource looks good.

**Comment by safern** [comment]
```
Applying: ActivityTagsCollection should implement IDictionary<string, object?>
/usr/bin/git push --force --set-upstream origin HEAD:backport/pr-41731-to-release/5.0-rc2
To https://github.com/dotnet/runtime
 + cc835beafa9...31405fa7873 HEAD -> backport/pr-41731-to-release/5.0-rc2 (forced update)
Branch 'backport/pr-41731-to-release/5.0-rc2' set up to track remote branch 'backport/pr-41731-to-release/5.0-rc2' from 'origin'.
Backport temp branch already exists, skipping opening a PR.
```

We need to re-open the PR.

@akoeplinger re-opening the PRs if the PR is not merged.

**Comment by safern** [comment]
I couldn't re-open the PR. So I deleted the branch and will run the backport again.

/backport to release/5.0-rc2

**Comment by krwq** [comment]
Changes from @buyaa-n look good to me (I cannot approve my own PR though)

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (roji) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: Sorry for the delayed response and thanks - yeah, this is wrong. GetSchemaTable should return nullable.

**Comment by roji** [comment]
> System.ComponentModel.Annotations by assuming we didn't annotated the src completely for a reason (not sure who annotated that, need to clarify)
> System.Data.Common same reason, not sure if we want to make the src and ref same

@buyaa-n @safern yeah, I'm the one who annotated System.ComponentModel.Annotations (#39611), and also System.Data.{Odbc,OleDb} (#39597). Unless I've missed something both that and System.Data.Common should be fully annotated (src and ref) apart from dependencies on TypeConverter and XML. IIRC in at least some of these (especially Odbc/OleDb) I couldn't get errors/warnings for discrepancies between src and ref, so there may have been mismatches...

Are there outstanding issues with these projects (sorry, am a bit late to the party..)? If so I can take a look - feel free to ping me etc.

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (roji) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs]: Yeah, I can confirm - when annotating System.Data.Common (and others) I deliberately left oblivious any API that depends on TypeConverters (as here) and XML. Those should be the only reason anything should still be oblivious in System.Data.Common (at least under ref).

**Comment by buyaa-n** [comment]
> Unless I've missed something both that and System.Data.Common should be fully annotated (src and ref) apart from dependencies on TypeConverter and XML. IIRC in at least some of these (especially Odbc/OleDb) I couldn't get errors/warnings for discrepancies between src and ref, so there may have been mismatches...

There is no error, we have run a tool to find diffs between ref and src and found some diff for those and other assemblies which we are trying to fix with this PR

> Are there outstanding issues with these projects (sorry, am a bit late to the party..)? If so I can take a look - feel free to ping me etc.

There is only one issue related to this where the list of findings attached https://github.com/dotnet/runtime/issues/41696.

**Comment by buyaa-n** [comment]
This must-have covered/fixed all diffs now (i will try to run @terrajobst's tool again), please review the fixes @safern @stephentoub @terrajobst

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs
Comment (jeffhandley) [inline src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:7]: Is it typical to add `using` statements in `ref` files?  I thought we left everything fully-qualified.

Comment (buyaa-n) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:7]: You are right, i will remove it

**Comment by buyaa-n** [comment]
Sorry 😥, it was not covered everything updated left over diffs in:
- `System.Data.OleDb` was not covered
- `System.IO.Packaging` - after enabling nullable most of the diff gone but still there were some diff left 

diff in `System.Resources.Extensions` were fixed after nullable enabled. Now it is covered all. Please review

**Review by jeffhandley — APPROVED** [review]
LGTM but please wait for other reviews.

File: src/libraries/System.Data.OleDb/src/OleDbTransaction.cs
Comment (buyaa-n) [inline src/libraries/System.Data.OleDb/src/OleDbTransaction.cs:140]: This is documented that could return null https://docs.microsoft.com/en-us/dotnet/api/system.data.oledb.oledbtransaction.connection?view=dotnet-plat-ext-3.1

Comment (roji) [reply src/libraries/System.Data.OleDb/src/OleDbTransaction.cs:140]: Yeah, the Connection property on all DbTransaction classes should indeed be nullable, thanks.

File: src/libraries/System.Security.Cryptography.Pkcs/ref/System.Security.Cryptography.Pkcs.netcoreapp.cs
Comment (safern) [reply src/libraries/System.Security.Cryptography.Pkcs/ref/System.Security.Cryptography.Pkcs.netcoreapp.cs:158]: I believe is because of the `NoWarn` we have on the ref files because we do: `throw null;`

**Review by safern — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs
Comment (roji) [inline src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs:133]: This method never seems to actually return null (even if GetSchemaTable in DbDataReader does), so we should probably annotate this as non-nullable.

Comment (buyaa-n) [reply src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs:133]: Yes, it was not returning null, and `?` was removed from ref, but reverted by review comment https://github.com/dotnet/runtime/pull/41731#discussion_r482841465, looked it again seems you are right, not returning even parent is nullable cc @stephentoub

**Comment by jeffhandley** [comment]
Updating the backport PR to capture latest commits...

/backport to release/5.0-rc2

File: src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs
Comment (jeffhandley) [reply src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs:133]: This one has gone back and forth and it's unclear to me what we're concluding is the final decision here. @buyaa-n / @roji Do we want to change this again to land on not nullable, or are we keeping it as nullable?

Comment (buyaa-n) [reply src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs:133]: As derived type may narrow the return type from `T?` to `T` and @roji confirmed it should be non null I am gonna update this to `T`

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs
Comment (jeffhandley) [reply src/libraries/System.Data.Common/ref/System.Data.Common.cs:866]: This was changed back per https://github.com/dotnet/runtime/pull/41731#discussion_r485268845

**Comment by jeffhandley** [comment]
Updating the backport PR to capture latest commit...

/backport to release/5.0-rc2

**Review by buyaa-n — APPROVED** [review]

File: src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs
Comment (krwq) [reply src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs]: if we're not sure we should annotate it as nullable so that it would not be a breaking change to change it back in vNext

Comment (roji) [reply src/libraries/System.Data.Common/src/System/Data/DataTableReader.cs]: @buyaa-n I think that's the right thing to do.

For context, DbDataReader.GetSchemaTable is indeed nullable, and null is returned when a database command doesn't return any resultset at all (or all resultsets have already been consumed on the reader via NextResult). This is how all ADO.NET providers behave (SqlClient, Odbc, OleDb...), see discussion in #41082 and #509.

DataTableReader is a specific implementation that exposes data in an existing DataTable via a DbDataReader. The implementation in this class never returns null, as a DataTable always provides an existing "resultset", so code working against DataTableReader can safely assume null will never be returned.

**Comment by jeffhandley** [comment]
One last backport to reflect this PR being merged.

/backport to release/5.0-rc2

**Pull request merged**
